src/span-data.ts

```diff
@@ -159,14 +179,16 @@ export class RootSpanData extends BaseSpanData implements RootSpan {
       return;
     }
     super.endSpan(timestamp);
-    traceWriter.get().writeTrace(this.trace);
-    this.children.forEach(child => {
-      if (!child.span.endTime) {
-        // Child hasn't ended yet.
-        // Inform the child that it needs to self-publish.
-        child.shouldSelfPublish = true;
-      }
-    });
+    if (this.traceWriter) {
```

Comment:
Consider the scenario where this code is reached, and `this.traceWriter` is not defined. Is that an error, or is that something very expected – and hence, silently ignorable, as being done here?

Reply:
Right now there's no code path that causes `this.traceWriter` not to be defined, but this code path will exist in a forthcoming PR, where we would opt to not publish a trace, but still propagate context.

Reply:
IMO it would be better for the code to be self-documenting. `if (this.shouldPublish) { /* publish code */ }` would be more obviously readable. Using the validness of `this.traceWriter` to accomplish that, but it doesn't clarify whether it is possible that some bug may cause failure to call `assignTraceWriter`, causing traces to be silently dropped.

IOW, my question is: did we really need to get rid of the singleton? Wouldn't it make sense to keep it a singleton, but add boolean state to each span on whether it should be published?

Reply:
> did we really need to get rid of the singleton?

I've decided that we don't need to get rid of it in order to address the issue at hand; I'll open a different PR with an alternative approach. That being said, I still think it is better not to depend on the singleton, and have the `TraceWriter` passed in -- it makes the code more testable and cleaner IMO. It should be considered a testable behavior of `Tracer` that it always calls `assignTraceWriter` so I don't know that a bug could crop up as long as we keep our tests intact. (I'll admit though that `assignTraceWriter` might not be the best name.)

Given that the change is ultimately inconsequential to users, though, I'd be OK with closing this PR.